internal/host/validator.go

```diff
@@ -658,4 +660,180 @@ func (v *validator) printSufficientOrUnknownInstallationDiskSpeed(c *validationC
 	default:
 		return fmt.Sprintf("Unexpected status %s", status.String())
 	}
+
+}
+
+func (v *validator) hasNetworkLatencyRequirementForRole(c *validationContext) ValidationStatus {
+
+	if len(c.cluster.Hosts) == 1 {
+		// Single Node use case
+		return ValidationSuccess
+	}
+	if len(c.host.Connectivity) == 0 {
+		return ValidationPending
+	}
+	latency, err := v.validateNetworkLatencyForRole(c.host, c.clusterHostRequirements, c.cluster.Hosts)
+	if err != nil || len(latency) > 0 {
+		return ValidationFailure
+	}
+	return ValidationSuccess
+}
+
+func getHostnameAndRoleByIP(ip string, hosts []*models.Host) (string, models.HostRole, error) {
+
+	for _, h := range hosts {
+		if h.Inventory != "" {
+			inv, err := hostutil.UnmarshalInventory(h.Inventory)
+			if err != nil {
+				return "", "", fmt.Errorf("unable to unmarshall cluster inventory for host %s", h.RequestedHostname)
+			}
+			if strings.Count(ip, ":") < 2 {
+				for _, i := range inv.Interfaces {
+					for _, ipv4 := range i.IPV4Addresses {
+						if ip == ipv4 {
+							return h.RequestedHostname, h.Role, nil
+						}
+					}
+				}
+			} else {
+				for _, i := range inv.Interfaces {
+					for _, ipv6 := range i.IPV6Addresses {
+						if ip == ipv6 {
+							return h.RequestedHostname, h.Role, nil
+						}
+					}
+				}
+			}
+		}
+	}
+	return "", "", fmt.Errorf("host with IP %s not found in inventory", ip)
+}
+
+func (v *validator) validateNetworkLatencyForRole(host *models.Host, clusterRoleReqs *models.ClusterHostRequirements, hosts []*models.Host) (map[string]string, error) {
+	var connectivityReport *models.ConnectivityReport
+	connectivityReport, err := hostutil.UnmarshalConnectivityReport(host.Connectivity)
+	if err != nil {
+		return nil, fmt.Errorf("unable to unmarshall host connectivity for %s:%s", host.ID, err)
+	}
+	failedHosts := map[string]string{}
+	for _, r := range connectivityReport.RemoteHosts {
+		for _, l3 := range r.L3Connectivity {
+			if l3.AverageRTTMs > clusterRoleReqs.Total.NetworkLatencyThresholdMs {
+				if _, ok := failedHosts[l3.RemoteIPAddress]; !ok {
+					hostname, role, err := getHostnameAndRoleByIP(l3.RemoteIPAddress, hosts)
+					if err != nil {
+						return nil, err
+					}
+					if role == host.Role {
+						failedHosts[l3.RemoteIPAddress] = hostname
+					}
+				}
+			}
+		}
+	}
+	return failedHosts, nil
+}
+
+func (v *validator) printHasNetworkLatencyRequirementForRole(c *validationContext, status ValidationStatus) string {
+	switch status {
+	case ValidationSuccess:
+		return "Network latency requirement have been satisfied."
+	case ValidationFailure:
+		hosts, err := v.validateNetworkLatencyForRole(c.host, c.clusterHostRequirements, c.cluster.Hosts)
```

Comment:
Too complicated.  If the function would return slice of IPs then you will be able to use this function https://golang.org/pkg/strings/#Join to join them into single string.
reflect is not needed here.

Reply:
I'll review it and refactor.